cloud.foundry.cli/src/main/java/cloud/foundry/cli/crosscutting/mapping/CfOperationsCreator.java

```diff
@@ -81,43 +77,26 @@ private static ReactorCloudFoundryClient createCloudFoundryClient(
                 .build();
     }
 
-    private static PasswordGrantTokenProvider createTokenProvider(LoginCommandOptions commandOptions)
-            throws Exception {
-
-        String user = determineCredential(commandOptions.getUserName(), CF_CONTROL_USER);
-        String password = determineCredential(commandOptions.getPassword(), CF_CONTROL_PASSWORD);
+    private static PasswordGrantTokenProvider createTokenProvider(LoginCommandOptions commandOptions) {
 
-        return PasswordGrantTokenProvider.builder()
-                .username(user)
-                .password(password)
-                .build();
-    }
+        String user = commandOptions.getUserName();
+        if (user == null) {
+            user = System.getenv(CF_CONTROL_USER);
+        }
 
-    /**
-     * Determines the credential-data, which is needed by the operations classes in the package.
-     * If the username and password is not set within the commandOptions,
-     * then it tries to get the values from the System Environment Variables:
-     * <b>"CF_CONTROL_USER" & "CF_CONTROL_PASSWORD"</b>
-     *
-     * @param credential Credential
-     * @param environmentVariableName Environment variable name
-     *
-     * @return Credential-Data
-     * @throws Exception
-     */
-    private static String determineCredential(String credential, String environmentVariableName) throws Exception {
-        String environmentVariableValue;
-        try {
-            environmentVariableValue = System.getenv(environmentVariableName);
-        } catch (SecurityException e) {
-            throw new CredentialException(SECURITY_EXCEPTION);
+        String password = commandOptions.getPassword();
+        if (password == null) {
+            password = System.getenv(CF_CONTROL_PASSWORD);
         }
 
-        if (credential == null && environmentVariableValue == null) {
-            throw new CredentialException(CREDENTIAL_INFORMATION_HAS_NOT_BEEN_DEFINED);
+        if (user == null || password == null) {
+            throw new MissingCredentialException(user, password);
```

Comment:
This exception is quite awkward to use. Took me a few minutes to wrap my head around it. Not sure if I'd come up with a better solution, though.

Reply:
Me neither :/ I'll leave it as it is for now...